File: leetcode/Python/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        sol, mx, mn = nums[0], nums[0], nums[0]
        for i in range(1, len(nums)):
            mx, mn = mx*nums[i], mn*nums[i]
            mx, mn = max(mx,mn,nums[i]), min(mx,mn,nums[i])
            sol = max(sol,mx)
        return sol

File: leetcode/C++/maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int sol=nums[0],mx=nums[0],mn=nums[0];
        for(int i=1,j=0,k=0; i<nums.size(); ++i){
            mx*=nums[i]; mn*=nums[i]; j=max(mx,mn); k=min(mx,mn);
            mx=max(j,nums[i]); mn=min(k,nums[i]);
            sol=max(sol,mx);
        }
        return sol;
    }
};